### ISSUE TITLE
Check of `_req` attribute in StreamResponse instance is useless

### ISSUE THREAD

**Issue by dizpers** [aio-libs/aiohttp#1554]
Hey there.

I use the code like the following to return the instance of StreamResponse:
```python
from aiohttp import web

async def handler(request):
    response = web.StreamResponse(status=200, reason='OK', headers={'Content-Type': 'text/html'})
    await response.prepare(request)
    # write data to response, call `drain` and so on
    return response
```

And I always get RuntimeError with message "Response has been started with different request". And it seems logical, because:

1. we initiate `_req` attribute with `None` value [in the constructor](https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web_reqrep.py#L518);
2. call of `prepare` method of StreamResponse instance [will lead to the call of `_start_pre_check` method](https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web_reqrep.py#L804);
3. and, finally, inside `_start_pre_check` method [we compare value of `_req` attribute with given request](https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web_reqrep.py#L769).

I mean that we don't fill `_req` attribute elsewhere, except [`_start` method](https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web_reqrep.py#L819), which is called only after the check (3).

The only workaround I see for now is to set that attribute manually like `response._req = request`. Could you tell me if such behaviour is expected or not.

Thanks.

**Comment by fafhrd91** [comment]
in step 3 we check `_req` only if `_resp_impl` is not None, but we set `_resp_impl` only in `_start` method.

could you create test case for your exception?

**Comment by dizpers** [comment]
Seems like I missing something. In some cases `_req` attribute contains expected request. I only have problems when cloning request like:

```python
async def handler(request):
    request = request.clone(...)
    response = web.StreamResponse(status=200, reason='OK', headers={'Content-Type': 'text/html'})
    await response.prepare(request)
    # write data to response, call `drain` and so on
    return response
```

You can see full example [here](https://github.com/dizpers/simple-mitm-proxy/blob/master/proxy.py#L11).

**Comment by fafhrd91** [comment]
you can not use cloned request for `prepare`, `web_server.RequestHandler` always calls prepare with original request.

**Comment by fafhrd91** [comment]
@asvetlov i think request cloning is confusing. we should add check to RequestHandler if response already prepared and do no call it again.

what was the reason for adding `clone` to `Request`?

**Comment by fafhrd91** [comment]
@dizpers let me think, we may need to prevent runtime exception for your case.

**Comment by dizpers** [comment]
Thank you for your quick help. :)

Initially request is going to local endpoint (like `127.0.0.1:8080`). I want to extract path from this request and get information from another host with the same path (simple reverse proxy behaviour). So, for new request I need:

1. path from original one;
2. modified headers (`Host`, `Referrer` - should be enough).

And, finally, I have to prepare response (StreamResponse in my case). Before writing data to it, I have to prepare it. And from what I've got recently - we need proper request object for this operation. So, I decided to use modified (by rules above) request object - by cloning it.

Do you think I can use original request for response preparation (I mean with wrong host and headers)?

**Comment by dizpers** [comment]
@fafhrd91 @asvetlov and actually `clone` method sounds for me like a good fit for this kind of task :)

**Comment by fafhrd91** [comment]
but you do not use cloned request, you just need request path and headers. just use original request.

from my understanding primary use case for cloned request is some kind of template rendering.
@asvetlov what was the reason for behind `clone` method?

**Comment by dizpers** [comment]
I can use original request and any extracted data to get information for the desired host (instead of local endpoint). But after I've got the desired data I have to return a response. And this response is the instance of StreamResponse, which should be preapared with some request. I guess that preparation process use information from the request (presented in arguments) to prepare proper response.

And I want that response to be formed properly (i.e. using right request with right data). Am I missing something?

**Comment by dizpers** [comment]
Cloned request will have only host and some headers modified. Is it important to build proper StreamResponse instance (prepare)?

**Comment by dizpers** [comment]
From what I see in the code - I guess I can use original request, as @fafhrd91 suggested.

**Comment by fafhrd91** [comment]
we meed request for response because we use low-level info from request and keep-alive header, other than that we don't need anything from request.

i will add extra check to request handler to prevent runtime exception.

**Comment by fafhrd91** [comment]
fixed in master by d2b3ecf150cbda5a1f965afa6f335e96e1c7550e

**Comment by dizpers** [comment]
Thank you very much!
